Fix navigation for creating a new pin by skipping current PIN or exceeding max retries

clauz9 opened the pull request:
### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I am following the [Code Style Guidelines](https://github.com/signalapp/Signal-Android/wiki/Code-Style-Guidelines)
- [x] I have tested my contribution on these devices:
 * Nokia 3.1, Android 10
- [x] My contribution is fully baked and ready to be merged as is
- [ ] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description
<!--
Describe briefly what your pull request proposes to fix. Especially if you have more than one commit, it is helpful to give a summary of what your contribution as a whole is trying to solve.
Also, please describe shortly how you tested that your fix actually works.
-->

There's an issue that happens because when you skip the current PIN and try to create a new one, the app navigates to `CreateKbsPinActivity` without finishing `PinRestoreActivity` allowing you to navigate back from the conversation list to `PinRestoreActivity`

<details>
<summary> GIF Recording of the issue </summary>

![before](https://user-images.githubusercontent.com/77681918/164188500-0b79c616-4959-4fdc-8992-c1a13b48617c.gif)

</details>


while looking into this issue, I've noticed there was [an older PR that was closed](https://github.com/signalapp/Signal-Android/pull/10314), that was dealing with this issue as well, so I pulled in the relevant commit,

and then added a new state in `PassphraseRequiredActivity` to prevent the app opening `EditProfileActivity` if you closed it while  `CreateKbsPinActivity` or "PinRestoredLockedFragment` was visible.

I've also disabled back press in PinRestoredLockedFragment.

clauz9 commented:
@signalien I hope it's ok that I pulled from your PR 🙂

clauz9 commented on app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java:
```diff
@@ -41,17 +41,18 @@ public abstract class PassphraseRequiredActivity extends BaseActivity implements
+  private static final int STATE_TRANSFER_ONGOING      = 8;
+  private static final int STATE_TRANSFER_LOCKED       = 9;
+  private static final int STATE_CHANGE_NUMBER_LOCK    = 10;
+  private static final int STATE_CREATE_NEW_SIGNAL_PIN = 11;
```
well 🤔 after tapping skip in `PinRestoreEntryFragment` 
`PinState.onPinRestoreForgottenOrSkipped()` is called which changes the state of the app 

https://github.com/signalapp/Signal-Android/blob/ff8f9ca81ae6a25e1e946612c817206b9410d9a1/app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreEntryFragment.java#L221-L224

https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/pin/PinState.java#L99-L104

and then you navigate to `CreateKbsPinActivity`, now if you close the app from here and then reopen it,
`PassphraseRequiredActivity.OnCreate()` is called which eventually calls `getApplicationState`
https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java#L163-L187

and that's where the state of the app is determined by those checks 🤔 and because we tapped Skip, 
https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/pin/PinState.java#L101


the state of the app is now determined to be `STATE_CREATE_PROFILE_NAME`
https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java#L172-L175

which launches `EditProfileActivity`

https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java#L155

https://github.com/signalapp/Signal-Android/blob/ed20c243267d9ea7acccfaea919b006d801e8f22/app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java#L234-L236

clauz9 replied:
Long story short 🙂  it prevents this from happening 

<details>
<summary> GIF Recording </summary>

![GIF 20 04 2022 11-26-17](https://user-images.githubusercontent.com/77681918/164513848-3005fa79-92a2-4632-8bb3-d5fd608b9189.gif)

</details>

which happens now because we finish the `PinRestoreActivity` in the current beta it goes back to `PinRestoreEntryFragment`

clauz9 commented on app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreActivity.java:
```diff
@@ -28,11 +28,12 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
     final Intent chained   = PassphraseRequiredActivity.chainIntent(createPin, main);
 
     startActivity(chained);
+    finish();
```
I don't pretend to be an expert in the Android SDK, but I assume it's because the `main` intent is chained with the `createPin` intent and that maybe means the Intent flags are not enforced 🤔 

I'm also looking at the `handleSuccess()` in `PinRestoreEntryFragment` that uses the same clearTop `main` intent and chains it with the `profile` intent and then still finishes the activity after the if/else 

https://github.com/signalapp/Signal-Android/blob/ff8f9ca81ae6a25e1e946612c817206b9410d9a1/app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreEntryFragment.java#L233-L252

clauz9 replied:
I'm still looking into this 🤔 

> I assume it's because the `main` intent is chained with the `createPin` intent and that maybe means the Intent flags are not enforced 🤔

I'm pretty sure this assumption is utterly wrong .. I can't find any documentation to support it... 🙂

clauz9 replied:
So... I finally did some homework on this, and I figured it out... 🤔 

From the [`FLAG_ACTIVITY_CLEAR_TOP` documentation](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_CLEAR_TOP) 

<details>
<summary>Clear Top Documentation</summary>

>If set, and the activity being launched is already running in the current task, then instead of launching a new instance of that activity, all of the other activities on top of it will be closed and this Intent will be delivered to the (now on top) old activity as a new Intent.
>
>For example, consider a task consisting of the activities: A, B, C, D. If D calls startActivity() with an Intent that resolves to the component of activity B, then C and D will be finished and B receive the given Intent, resulting in the stack now being: A, B.
>
>The currently running instance of activity B in the above example will either receive the new intent you are starting here in its onNewIntent() method, or be itself finished and restarted with the new intent. If it has declared its launch mode to be "multiple" (the default) and you have not set [FLAG_ACTIVITY_SINGLE_TOP](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_SINGLE_TOP) in the same intent, then it will be finished and re-created; for all other launch modes or if [FLAG_ACTIVITY_SINGLE_TOP](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_SINGLE_TOP) is set then this Intent will be delivered to the current instance's onNewIntent().
>
>This launch mode can also be used to good effect in conjunction with [FLAG_ACTIVITY_NEW_TASK](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_NEW_TASK): if used to start the root activity of a task, it will bring any currently running instance of that task to the foreground, and then clear it to its root state. This is especially useful, for example, when launching an activity from the notification manager.

</details>

now our task will contain the following activities when starting the app, and registering with a phone number, and skipping `PinRestoreActivity`

Task 1:
{ `PassphraseRequiredActivity`(A) -> `RegistrationNavigationActivity`(B) -> `PinRestoreActivity`(C) 

now when we tap `skip` in `PinRestoreActivity` `navigateToPinCreation()` is called,

https://github.com/signalapp/Signal-Android/blob/38836198a1813c8d7c02d7c68feea8fbcb00bb19/app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreActivity.java#L31-L37

which calls `clearTop()` from `MainActivity`

https://github.com/signalapp/Signal-Android/blob/38836198a1813c8d7c02d7c68feea8fbcb00bb19/app/src/main/java/org/thoughtcrime/securesms/MainActivity.java#L40-L48

then chains `CreateKbsPinActivity` with `MainActivity` and starts them. 

`CreateKbsPinActivity` will become activity (D) in Task 1, and when `MainActivity` is started, it will become activity (A) in Task 2

so when we reach the conversation list our back stack looks like this..

Task 1: { `PassphraseRequiredActivity`(A) -> `RegistrationNavigationActivity`(B) -> `PinRestoreActivity`(C) -> `CreateKbsActivity`(D) } 
Task 2: {`MainActivity`(A) -> .... }

now according to the documentation for `CLEAR_TOP`, if `MainActivity` would have been activity (A) of Task 1 and `PinRestoreActivity`(C) would have started an intent to `MainActivity` then yes all other activities in Task 1 would have been finished, 

but `MainActivity` has never been started before, so it just starts normally in a new task, and leaves whatever existing activities in Task 1 unfinished. 

so we end up with two options, 🤔  we either use `activity.finish()` to finish `PinRestoreActivity` or we can use an intent for `MainActivity` with the following flags. 

`Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK`

which according to the [documentation for the `CLEAR_TASK` flag](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_CLEAR_TASK) 

<details>
<summary>Clear Task Documentation</summary>

>If set in an Intent passed to [Context.startActivity()](https://developer.android.com/reference/android/content/Context#startActivity(android.content.Intent)), this flag will cause any existing task that would be associated with the activity to be cleared before the activity is started. That is, the activity becomes the new root of an otherwise empty task, and any old activities are finished. This can only be used in conjunction with [FLAG_ACTIVITY_NEW_TASK](https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_NEW_TASK).

</details>

`CLEAR_TASK` & `NEW_TASK` will finish all activities in Task 1 and make `MainActivity` the root activity.

Now, I tested using `CLEAR_TASK` & `NEW_TASK` and it works, but since `MainActivity.clearTop()` is used in a lot of other places, we probably shouldn't change it. 

so adding `activity.finish()` to `PinRestoreActivity.navigateToPinCreation()` is the simpler approach. 

the rest of the code changes to ensure the app state is determined correctly, would still be necessary no matter what approach we take to finish the activity.

clauz9 commented on app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreActivity.java:
```diff
@@ -28,11 +28,12 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
   }
 
-  void navigateToPinCreation() {
+  void navigateToPinCreationAndFinish() {
```
I figured, it's good practice to know what a method is doing by it's name, so I renamed it, to reflect the fact that it will finish the activity, since it doesn't currently do that. 🤔 

in any case I made it a sperate commit just in case it's considered not necessary, so it can be left out 👍

clauz9 commented on app/src/main/java/org/thoughtcrime/securesms/pin/PinRestoreLockedFragment.java:
```diff
@@ -26,11 +27,23 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
       CommunicationActions.openBrowserLink(requireContext(), getString(R.string.PinRestoreLockedFragment_learn_more_url));
     });
+
+    disableOnBackPressed();
```
I guess you're right, I'll try doing what you suggested 🤔 

Ideally I think the app should reopen on the same `PinRestoreLockedFragment` after exiting, so the user can read the info on that screen

but it will not, it will either open in `CreateKbsPinActivity` if we use the new app state that I added, 
because I made sure to change the app state slightly earlier in 

https://github.com/signalapp/Signal-Android/pull/12183/commits/97c8c8b98752205f231ad7e78ac86631683d450a#diff-acdb1656867683a17db852afe616c4cc28ec66685701c323baa876454d219ee5R229-R232

or in `EditProfileActivity` if we don't use the new state

clauz9 commented:
I reverted the "rename..." commit, and the "prevent backing out..." commit, and changed the nav graph to pop up to root inclusive, which will close the app when you press back 

Also I merged in the most recent version

cody-signal commented on app/src/main/java/org/thoughtcrime/securesms/PassphraseRequiredActivity.java:
```diff
@@ -219,6 +227,10 @@ private Intent getEnterSignalPinIntent() {
     return getRoutedIntent(PinRestoreActivity.class, getIntent());
   }
 
+  private Intent getCreateNewSignalPinIntent() {
```
Why do we need a second one instead of using the existing, I assume if we are making a new pin we'll also need to make new profile info unless the create flow takes you to profile afterwards which it probably does in the reg complete fragment.

cody-signal reviewed (commented):
One follow up question, and then please squash into a single commit.

clauz9 commented:
So I did both, I've moved the `userMustCreateSignalPin()` above `userMustSetProfileName()` in this commit https://github.com/signalapp/Signal-Android/pull/12183/commits/e15e67eda79375992eccfa04b3fc5ea0c755b787

and added a `pinForgottenOrSkippedKey` in `KbsValues` in this commit https://github.com/signalapp/Signal-Android/pull/12183/commits/7e4c75ea4bf5144a5c671979688c43c49ea5ca3c 

I've also rebased it on the most recent version, and squashed all the previous commits. 

this commit https://github.com/signalapp/Signal-Android/pull/12183/commits/7e4c75ea4bf5144a5c671979688c43c49ea5ca3c is not fully tested, I'll do it tomorrow.

clauz9 commented:
so.. I tested this, https://github.com/signalapp/Signal-Android/pull/12183/commits/e15e67eda79375992eccfa04b3fc5ea0c755b787, and this https://github.com/signalapp/Signal-Android/pull/12183/commits/d6543e98b6db263e81ac950cb5b41df1416cc674 

In the following scenarios, a clean registration scenario, a pin restore scenario, a skip scenario, and an unregister - reregister scenario.. everything seems to be ok.

cody-signal commented:
Please squash and I'll bring it in.

clauz9 commented:
> Please squash and I'll bring it in.

Done, @cody-signal sorry for the delay, had some things to do today, that kept me away from a PC 🙂

clauz9 commented:
This didn't make it into 5.37 🤔
